### PR TITLE
Harmony/unify component graphs

### PR DESCRIPTION
**Pull request by lalybar**
## Proposed Changes

-  Added graph extension
-  Added insights extension
-  Updated cleargraph version to 5.2.0 (latest)

---
This change is [Reviewable](https://reviewable.io/reviews/teambit/bit/2470)

### CODE REVIEW

File: src/extensions/insights/insights.cmd.tsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      const listItems = results.map(insight => (insight += '\n'));
+      return <Color blueBright>{listItems}</Color>;
+    }
+    if (names) {
+      let results: InsightResult[] = [];
+      if (Array.isArray(names)) {
+        results = await this.insightManager.run(names);
+      } else {
+        results = await this.insightManager.run([names]);
+      }
```
Comment (GiladShoham) [inline src/extensions/insights/insights.cmd.tsx]: would be more elegant to do something like
```js
const namesArr = (typeof names === 'string') ? [names] : names
const results = await this.insightManager.run(namesArr);
```

File: src/extensions/insights/insights.cmd.tsx [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      return <Color blueBright>{results}</Color>;
+    } else {
```
Comment (GiladShoham) [inline src/extensions/insights/insights.cmd.tsx]: remove the else, you don't need else after return

Comment (GiladShoham) [reply src/extensions/insights/insights.cmd.tsx]: Acutally you will need it once you won't return here but rather take the results into the react component

File: src/extensions/insights/insights.cmd.tsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                    {result.metaData.name}
+                    {'\n'}
+                  </Text>
+                </Box>
+                <Text>{'\n'}</Text>
+                <Box>{result.renderedData}</Box>
+              </Box>
+            );
+          })}
+        </Box>
```
Comment (GiladShoham) [inline src/extensions/insights/insights.cmd.tsx]: I would move this into another function and do something like
```
results = names ? await this.insightManager.run([names]) : await this.insightManager.runAll();
return template(results)
function template(results) : {
<Box key="help" flexDirection="column">
//....
</Box>
}
```

File: src/extensions/insights/insight.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  metaData: InsightMetaData;
+  message?: string;
+  data: any;
+  renderedData: any; // TOD) - should be react component
+};
+
+export type RawResult = {
+  message?: string;
+  data: any;
+};
```
Comment (GiladShoham) [inline src/extensions/insights/insight.ts]: I don't think you need it. it's something internal for each insight and not must.

File: src/extensions/insights/insight.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+export type InsightMetaData = {
+  name: string;
+  description: string;
+};
+
+export type InsightResult = {
+  metaData: InsightMetaData;
+  message?: string;
+  data: any;
+  renderedData: any; // TOD) - should be react component
```
Comment (GiladShoham) [inline src/extensions/insights/insight.ts]: React.ReactElement

File: src/extensions/insights/insight-manager.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * execute an array of insights
+   *
+   */
+  async run(insightNames: string[]): Promise<InsightResult[]> {
+    const res: InsightResult[] = [];
+    await Promise.all(
+      insightNames.map(async insightName => {
+        const insight = this.getByName(insightName);
+        if (!!insight) {
```
Comment (GiladShoham) [inline src/extensions/insights/insight-manager.ts]: you can change to `if (insight) {`

File: src/extensions/insights/insight-manager.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    return res;
+  }
+
+  /**
+   * execute all insights in the registry
+   *
+   */
+  async runAll(): Promise<InsightResult[]> {
+    const res: InsightResult[] = [];
+    for (let [name, insight] of this.insights.entries()) {
```
Comment (GiladShoham) [inline src/extensions/insights/insight-manager.ts]: Any reason why not run them in parallel?

File: src/extensions/insights/all-insights/find-cycles.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    this.graph = graph;
+  }
+  async _runInsight(): Promise<RawResult> {
+    const cycles = this.graph.findCycles();
+    if (cycles.length === 1) {
+      return {
+        message: `Found ${cycles.length} cycle.`,
+        data: cycles
+      };
+    } else {
```
Comment (GiladShoham) [inline src/extensions/insights/all-insights/find-cycles.tsx]: remove the else

File: src/extensions/insights/all-insights/find-cycles.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    const result = {
+      metaData: {
+        name: this.name,
+        description: this.description
+      },
+      data: bareResult.data,
+      renderedData: renderedData
+    };
+
+    if (!!bareResult.message) {
```
Comment (GiladShoham) [inline src/extensions/insights/all-insights/find-cycles.tsx]: you can use `if (bareResult.message) {`

File: src/extensions/insights/all-insights/duplicate-dependencies.tsx [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+      formatted.push({
+        dependencyId: dependency,
+        latestVersion: depData.latestVersionId,
+        dependents: dependents
+      })
+    }
+    return formatted;
+  }
+
+  _getDependents(priorVersions: VersionSubgraph[]): {id: string, usedVersion: string}[]{
```
Comment (GiladShoham) [inline src/extensions/insights/all-insights/duplicate-dependencies.tsx]: Make the `{id: string, usedVersion: string}` a declared type (you are using it twice)

File: src/extensions/insights/all-insights/duplicate-dependencies.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  return <Box>
+    <Color grey>{m}</Color>
+  </Box>
+}
+function alignCommandName(name: string, sizeToAlign = 20) {
+  return `${name}${new Array(sizeToAlign).join(' ')}`;
+}
+
+
+
```
Comment (GiladShoham) [inline src/extensions/insights/all-insights/duplicate-dependencies.tsx]: remove

File: src/consumer/component/component-loader.ts [hunk 12]
```diff
@@ -143,6 +143,7 @@ export default class ComponentLoader {
       );
       addExtensionsAsDevDependencies(component);
       updateDependenciesVersions(this.consumer, component);
+      addExtensionsAsDevDependencies(component);
```
Comment (GiladShoham) [inline src/consumer/component/component-loader.ts]: why did you add this? is it by mistake?

File: src/extensions/component/component-factory.ts [hunk 13]
```diff
@@ -14,10 +17,12 @@ export default class ComponentFactory {
 
   create() {}
 
+  transformToComponents(workspaceComponents: ConsumerComponent[], modelComponents: ModelComponent[]) {}
+
   /**
    * instantiate a component object from a legacy `ComponentVersions` type object.
    */
-  fromComponentVersions() {}
+  fromComponentVersions(legacyComponent: ModelComponent, version: Version) {}
```
Comment (GiladShoham) [inline src/extensions/component/component-factory.ts]: do you really need these changes?

File: src/extensions/graph/component-graph/component-graph.ts [hunk 14]
```diff
@@ -1,9 +1,120 @@
-import { Graph } from 'cleargraph';
+import { Graph, NodeDoesntExist } from 'cleargraph';
```
Comment (GiladShoham) [inline src/extensions/graph/component-graph/component-graph.ts]: NodeDoesntExist not in use

File: src/extensions/graph/component-graph/component-graph.ts [hunk 15]
```diff
@@ -1,9 +1,120 @@
...
+  versionMap: Map<string, { allVersionNodes: string[]; latestVersionNode: string }>;
+  constructor(
+    nodes: { id: string; node: Component }[] = [],
+    edges: { sourceId: string; targetId: string; edge: Dependency }[] = []
+  ) {
+    super(nodes, edges);
+    this.versionMap = new Map();
+  }
+  static buildFromLegacy(legacyGraph: LegacyGraph, componentFactory: ComponentFactory): Graph<Component, Dependency> {
+    let newGraph = new ComponentGraph();
```
Comment (GiladShoham) [inline src/extensions/graph/component-graph/component-graph.ts]: const

File: src/extensions/graph/component-graph/component-graph.ts [hunk 16]
```diff
@@ -1,9 +1,120 @@
...
 
-// TODO: add concrete types
-export default class ComponentGraph extends Graph<Component, Dependency> {}
+export class ComponentGraph extends Graph<Component, Dependency> {
+  versionMap: Map<string, { allVersionNodes: string[]; latestVersionNode: string }>;
+  constructor(
+    nodes: { id: string; node: Component }[] = [],
+    edges: { sourceId: string; targetId: string; edge: Dependency }[] = []
+  ) {
+    super(nodes, edges);
```
Comment (GiladShoham) [inline src/extensions/graph/component-graph/component-graph.ts]: type issue

File: src/extensions/component/component.ts [hunk 17]
```diff
@@ -52,6 +53,14 @@ export default class Component {
     return this.state.filesystem;
   }
 
+  toString(): string {
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/extensions/component/component.ts]: First, I renamed it stringify() because this is what cleargraph requires. Second, not sure what needs to be considered - what it will print?

File: src/extensions/graph/dependency.ts [hunk 18]
```diff
@@ -0,0 +1,11 @@
+type DependencyTypes = 'dev' | 'peer' | 'runtime';
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/extensions/graph/dependency.ts]: Correct

File: src/extensions/graph/graph-api.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+export interface command {
+  inisight: insightName;
+  graph: Graph<Component, Dependency>;
+  startingNode?: nodeId;
+  depth?: number; // the max number of steps for traversal
+}
+
+export interface insights {
+  data: Array<insight>;
+}
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/extensions/graph/graph-api.ts]: deleted

File: src/extensions/component/component-factory.ts (hunk 13 above)
Comment (lalybar) [reply src/extensions/component/component-factory.ts]: deleted

File: src/extensions/graph/graph.manifest.ts [hunk 20]
```diff
@@ -1,10 +1,10 @@
...
+import { provide } from './graph.provider';
 import { WorkspaceExt } from '../workspace';
+import { ComponentFactoryExt } from '../component';
 
 export default {
   name: 'graph',
-  dependencies: [ComponentFactoryExt, BitCliExt, WorkspaceExt],
-  provider: Graph.provide
+  dependencies: [WorkspaceExt, ComponentFactoryExt],
+  config: {},
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/extensions/graph/graph.manifest.ts]: removed

File: src/extensions/insights/exceptions/insight-already-exists.ts [hunk 21]
```diff
@@ -0,0 +1,10 @@
+import AbstractError from '../../../error/abstract-error';
+
+export default class InsightAlreadyExists extends AbstractError {
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/extensions/insights/exceptions/insight-already-exists.ts]: It requires implementing a render method. Are we sure we want to add places that use Ink?

File: src/extensions/insights/insight.provider.ts [hunk 22]
```diff
@@ -0,0 +1,22 @@
+import { ComponentGraph } from '../graph/component-graph';
+import getCoreInsights from './core-insights-getter';
+import { BitCli } from '../cli';
+import { InsightManager } from './insight-manager';
+import { Insight } from './insight';
+import InsightsCmd from './insights.cmd';
+
+export type InsightConfig = {
+  silence: boolean;
+};
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/extensions/insights/insight.provider.ts]: removed

File: src/version/version.ts [hunk 23]
```diff
@@ -30,4 +30,8 @@ export default class Version {
     if (this.versionNum && !this.latest) return this.versionNum.toString();
     throw new InvalidVersion(this.versionNum);
   }
+
+  isLaterThan(otherVersion: Version) {
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/version/version.ts]: Left as isLaterThan because of getLatest and because it doesn't only wrap semver.gt() but also addresses "latest"

File: src/extensions/insights/insights.cmd.tsx [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+      );
+    }
+  }
+
+  // async someRender([components]: CLIArgs, { verbose, noCache }: Flags) {
+  //   // @ts-ignore
+  //   const compileResults = await this.compile.compile(components, { verbose, noCache });
+  //   // eslint-disable-next-line no-console
```
[the remark this answers is not in the document]
Comment (lalybar) [reply src/extensions/insights/insights.cmd.tsx]: Oops :)

File: src/extensions/insights/insights.cmd.tsx (hunk 1 above)
Comment (lalybar) [reply src/extensions/insights/insights.cmd.tsx]: Corrected (except for the const before results.

File: src/extensions/insights/insights.cmd.tsx (hunk 3 above)
Comment (lalybar) [reply src/extensions/insights/insights.cmd.tsx]: done

File: src/extensions/insights/insight.ts (hunk 4 above)
Comment (lalybar) [reply src/extensions/insights/insight.ts]: I find it very convenient to have it there. Insight doesn't force the user to use it. It's there as a reference.

File: src/extensions/insights/insight.ts (hunk 5 above)
Comment (lalybar) [reply src/extensions/insights/insight.ts]: thanks :)

File: src/extensions/insights/insight-manager.ts (hunk 6 above)
Comment (lalybar) [reply src/extensions/insights/insight-manager.ts]: isn't !! best practice in these cases?

Comment (lalybar) [reply src/extensions/insights/insight-manager.ts]: And if not - why does it exist?

File: src/extensions/insights/insight-manager.ts (hunk 7 above)
Comment (lalybar) [reply src/extensions/insights/insight-manager.ts]: No reason. Currently there are no dependencies between them.
Not sure how to do that though, and if it's a good idea going forward, in case there are some dependencies in the future.

File: src/extensions/insights/all-insights/find-cycles.tsx (hunk 9 above)
Comment (lalybar) [reply src/extensions/insights/all-insights/find-cycles.tsx]: see my comment about !! before.

File: src/extensions/insights/all-insights/duplicate-dependencies.tsx (hunk 10 above)
Comment (lalybar) [reply src/extensions/insights/all-insights/duplicate-dependencies.tsx]: fixed

File: src/extensions/insights/all-insights/duplicate-dependencies.tsx (hunk 11 above)
Comment (lalybar) [reply src/extensions/insights/all-insights/duplicate-dependencies.tsx]: removed

File: src/consumer/component/component-loader.ts (hunk 12 above)
Comment (lalybar) [reply src/consumer/component/component-loader.ts]: Not sure. Might have been in one of my conversations with David about the graph. Awaiting his response.

File: src/extensions/graph/component-graph/component-graph.ts (hunk 16 above)
Comment (lalybar) [reply src/extensions/graph/component-graph/component-graph.ts]: fixed
